src/Symfony/Component/Serializer/Normalizer/NonBackedEnumNormalizer.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\PropertyInfo\Type;
+use Symfony\Component\Serializer\Exception\InvalidArgumentException;
+use Symfony\Component\Serializer\Exception\NotNormalizableValueException;
+
+/**
+ * Normalizes a {@see \UnitEnum} that is not a {@see \BackedEnum} to a string identifier.
+ *
+ * @author Misha Kulakovsky <misha@kulakovs.ky>
+ */
+final class NonBackedEnumNormalizer implements NormalizerInterface, DenormalizerInterface
+{
+    public function getSupportedTypes(?string $format): array
+    {
+        return [
+            \UnitEnum::class => true,
+        ];
+    }
+
+    public function normalize(mixed $object, string $format = null, array $context = []): int|string
+    {
+        if (!$object instanceof \UnitEnum || $object instanceof \BackedEnum) {
+            throw new InvalidArgumentException('The data must belong to a non-backed enumeration.');
+        }
+
+        return $object->name;
+    }
+
+    public function supportsNormalization(mixed $data, string $format = null, array $context = []): bool
+    {
+        return $data instanceof \UnitEnum
+            && !$data instanceof \BackedEnum;
+    }
+
+    /**
+     * @throws NotNormalizableValueException
+     */
+    public function denormalize(mixed $data, string $type, string $format = null, array $context = []): mixed
+    {
+        if (!is_subclass_of($type, \UnitEnum::class) || is_subclass_of($type, \BackedEnum::class)) {
+            throw new InvalidArgumentException('The data must belong to a non-backed enumeration.');
+        }
+
+        if (!\is_string($data) && !($context[BackedEnumNormalizer::ALLOW_INVALID_VALUES] ?? false)) {
+            throw NotNormalizableValueException::createForUnexpectedDataType("The data is not a string, you should pass a string that can be parsed as an enumeration case of type $type.", $data, [Type::BUILTIN_TYPE_STRING], $context['deserialization_path'] ?? null, true);
+        }
+
+        try {
+            $constantName = "$type::$data";
+            if (!\defined($constantName)) {
+                throw new \ValueError("$data is not a valid case for enum $type.");
+            }
+
+            $value = \constant($constantName);
+            if (!\is_object($value) || $value::class !== $type) {
```

Comment:
Could you elaborate a bit more on under what scenario this condition can reach `true` at this point?

Reply:
Using `constant(..)` is a way to address enum cases by their fqcn+name, but enums can have real constants too:

```php
enum MyEnum: string {
    const A = "a";
    case B = "b";
}

var_dump(constant('MyEnum::A')); // string(1) "a"
var_dump(constant('MyEnum::B')); // enum(MyEnum::B)
```

This is why we should check again that data was an enum reference.